Bottom: 1144d68132e2342887591758840329ab3f4487d6
Top:    060400e1ef5c86f53c7a1e5e245ad777514a9ff8
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2016-12-04 19:24:18 +0000

Fix labeling of contour which ends up as all SHARED / SHARED2, or all UNKNKWN

SHARED / SHARED2 labelling does not propagate, so if this occurs for an entire
contour, the labelling code would loop forever. (Same as if no cross-connected
vertex with definitive INSIDE / OUTSIDE information was found).

Catch when we wrap around the edge list and then decide how to classify the
whole contour. If it is all UNKNWN, we re-classify it as not intersected. The
cntr_label_POLYAREA() routine is adjusted to handle the possibility that the
label_contour() routine for ISECTED contours changes the contour status back
back to UNKNWN.

If the whole contour is SHARED or SHARED2, we leave it as ISECTED, since there
is no handling of that condition for an entire contour. (It will either be
dropped, or re-constructed, depending on the boolean operation being performed).


---

diff --git a/src/polygon1.c b/src/polygon1.c
index b726d7b..0c7b78c 100644
--- a/src/polygon1.c
+++ b/src/polygon1.c
@@ -1218,7 +1218,24 @@ label_contour (PLINE * a)
       assert (label == INSIDE || label == OUTSIDE);
       LABEL_EDGE (cure, label);
     }
-  while ((cure = NEXT_EDGE (cure)) != first_labelled);
+  while ((cure = NEXT_EDGE (cure)) != first_labelled && (first_labelled != NULL || cure != &a->head));
+
+  if (first_labelled == NULL)
+    {
+      if (EDGE_LABEL (&a->head) == UNKNWN)
+        {
+          g_warning ("Walked entire contour and couldn't find anything we could label - it is either all INSIDE or OUTSIDE");
+          /* Mark the contour as NOT intersected, so it can be treated separately below. */
+          /* XXX: Does this work with separated out intersected contours? */
+          a->Flags.status = UNKNWN;
+        }
+      else
+        {
+          g_warning ("Walked entire contour and couldn't find anything we could label - it is either all SHARED OR SHARED2");
+          /* Head was marked, so presumably the entire contour is either SHARED or SHARED2 */
+        }
+    }
+
 #warning The above loop could run forever if we encounter a contour where the only intersection gets nuked due to shared edges
 #ifdef DEBUG_ALL_LABELS
   print_labels (a);
@@ -1235,7 +1252,12 @@ cntr_label_POLYAREA (PLINE * poly, POLYAREA * ppl, BOOLp test)
     {
       label_contour (poly);	/* should never get here when BOOLp is true */
     }
-  else if (cntr_in_M_POLYAREA (poly, ppl, test))
+
+  /* label_contour may decide the contour is NOT intersected, and we must fall through to the tests below */
+  if (poly->Flags.status == ISECTED)
+    return false;
+
+  if (cntr_in_M_POLYAREA (poly, ppl, test))
     {
       if (test)
 	return TRUE;
@@ -2328,6 +2350,7 @@ M_POLYAREA_Collect_separated (jmp_buf * e, PLINE * afst, POLYAREA ** contours,
       if (cntr_Collect (e, cur, contours, holes, action, NULL, NULL, NULL))
 	next = cur;
     }
+  /* XXX: What about rogue contours we re-labelled as INSIDE or OUTSIDE - ARE THEY DEALT WITH OK? */
 }
 
 static void
